src/fcntl.rs

```diff
@@ -185,28 +185,70 @@ fn wrap_readlink_result(v: &mut Vec<u8>, res: ssize_t) -> Result<OsString> {
         Err(err) => Err(err),
         Ok(len) => {
             unsafe { v.set_len(len as usize) }
+            v.shrink_to_fit();
             Ok(OsString::from_vec(v.to_vec()))
         }
     }
 }
 
-pub fn readlink<P: ?Sized + NixPath>(path: &P) -> Result<OsString> {
+fn inner_readlink<P: ?Sized + NixPath>(dirfd: Option<RawFd>, path: &P) -> Result<OsString> {
     let mut v = Vec::with_capacity(libc::PATH_MAX as usize);
+    // simple case: result is strictly less than `PATH_MAX`
     let res = path.with_nix_path(|cstr| {
-        unsafe { libc::readlink(cstr.as_ptr(), v.as_mut_ptr() as *mut c_char, v.capacity() as size_t) }
+        unsafe {
+            match dirfd  {
+                Some(dirfd) => libc::readlinkat(dirfd, cstr.as_ptr(), v.as_mut_ptr() as *mut c_char, v.capacity() as size_t),
+                None => libc::readlink(cstr.as_ptr(), v.as_mut_ptr() as *mut c_char, v.capacity() as size_t),
+            }
+        }
     })?;
+    let len = Errno::result(res)?;
+    debug_assert!(len >= 0);
+    if (len as usize) < v.capacity() {
+        return wrap_readlink_result(&mut v, res);
+    }
+    // Uh oh, the result is too long...
+    // Let's try to ask lstat how many bytes to allocate.
+    let reported_size = super::sys::stat::lstat(path).and_then(|x| Ok(x.st_size)).unwrap_or(0);
+    let mut try_size = if reported_size > 0 { reported_size as usize + 1 }
+    // If lstat doesn't cooperate, or reports an error, be a little less precise.
+    else { (libc::PATH_MAX as usize).max(128) << 1 };
+    loop {
+        v.reserve_exact(try_size);
+        let res = path.with_nix_path(|cstr| {
+            unsafe {
+                match dirfd {
+                    Some(dirfd) => libc::readlinkat(dirfd, cstr.as_ptr(), v.as_mut_ptr() as *mut c_char, v.capacity() as size_t),
+                    None => libc::readlink(cstr.as_ptr(), v.as_mut_ptr() as *mut c_char, v.capacity() as size_t),
+                }
+            }
+        })?;
+        let len = Errno::result(res)?;
+        debug_assert!(len >= 0);
+        if (len as usize) < v.capacity() {
+            break wrap_readlink_result(&mut v, res);
+        }
+        else {
+            // Ugh! Still not big enough!
+            let next_size = try_size << 1;
```

Comment:
I don't think we need to gracefully handle symlinks with paths greater than 32 bits.  A panic would suffice.  You can do
```rust
try_size = try_size.checked_mul(2).unwrap();
```

Reply:
I think it's better to return an error in this case (as below), but if you ask me to I'll replace that logic with a checked multiply.

Reply:
In fact, I think this code will still panic if the size overflows 32 bits.  I think `<<` , like most operators, will panic on overflow in debug builds (but wrap on release builds).  If you really want to handle the overflow without panicing, you need to use `checked_shl`.